File: GFG/DP/42.Printing-Longest-Increasing-Subsequence.cpp
// Printing Longest Increasing Subsequence

class Solution {
  public:
    vector<int> longestIncreasingSubsequence(int n, vector<int>& arr) {
        // Code here
        vector<int> dp(n,1), hash(n);
        int lastInd = 0;
        for(int i=0; i<n; i++) {
            hash[i] = i;
            for(int prev=0; prev<=i-1; prev++) {
                if(arr[i]>arr[prev] && 1+dp[prev]>dp[i]) {
                    dp[i] = 1 + dp[prev];
                    hash[i] = prev;
                }
            }
            if(dp[i] > dp[lastInd]) lastInd = i;
        }
        
        vector<int> ans;
        ans.push_back(arr[lastInd]);
        while(hash[lastInd] != lastInd) {
            lastInd = hash[lastInd];
            ans.push_back(arr[lastInd]);
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};

File: GFG/DP/42.Printing-Longest-Increasing-Subsequence.java
// Printing Longest Increasing Subsequence

class Solution {
    public ArrayList<Integer> getLIS(int nums[]) {
        // Code here
        int n = nums.length;
        
        int[] dp = new int[n];
        int[] prev = new int[n];
        for (int i = 0; i < n; i++) {
            dp[i] = 1;
            prev[i] = i;
        }
        
        int lastInd = -1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j] && 1 + dp[j] > dp[i]) {
                    dp[i] = 1 + dp[j];
                    prev[i] = j;
                }
            }
            if (lastInd == -1 || dp[i] > dp[lastInd]) {
                lastInd = i;
            }
        }
        
        ArrayList<Integer> ans = new ArrayList<>();
        while (prev[lastInd] != lastInd) {
            ans.add(nums[lastInd]);
            lastInd = prev[lastInd];
        }
        ans.add(nums[lastInd]);
        
        Collections.reverse(ans);
        return ans;
    }
}
